Editors/RoomEditor.cpp

```diff
@@ -1,13 +1,141 @@
 
-  QGraphicsScene* scene = new QGraphicsScene(this);
-  ui->graphicsView->setScene(scene);
+  ProtoModel* roomModel = model->GetSubModel(TreeNode::kRoomFieldNumber);
+  ui->roomRenderer->SetResourceModel(roomModel);
+
+  nodeMapper->addMapping(ui->roomName, TreeNode::kNameFieldNumber);
+  nodeMapper->toFirst();
+
+  resMapper->addMapping(ui->speedSpinBox, Room::kSpeedFieldNumber);
+  resMapper->addMapping(ui->widthSpinBox, Room::kWidthFieldNumber);
+  resMapper->addMapping(ui->heightSpinBox, Room::kHeightFieldNumber);
+  resMapper->addMapping(ui->clearCheckBox, Room::kClearDisplayBufferFieldNumber);
+  resMapper->addMapping(ui->persistentCheckBox, Room::kPersistentFieldNumber);
+  resMapper->addMapping(ui->captionLineEdit, Room::kCaptionFieldNumber);
+
+  resMapper->addMapping(ui->enableViewsCheckBox, Room::kEnableViewsFieldNumber);
+  resMapper->addMapping(ui->clearViewportCheckBox, Room::kClearViewBackgroundFieldNumber);
+  resMapper->toFirst();
+
+  RepeatedProtoModel* vm = roomModel->GetRepeatedSubModel(Room::kViewsFieldNumber);
+  ImmediateDataWidgetMapper* viewMapper = new ImmediateDataWidgetMapper(this);
+  viewMapper->setModel(vm);
+  viewMapper->addMapping(ui->viewVisibleCheckBox, View::kVisibleFieldNumber);
+
+  viewMapper->addMapping(ui->cameraXSpinBox, View::kXviewFieldNumber);
+  viewMapper->addMapping(ui->cameraYSpinBox, View::kYviewFieldNumber);
+  viewMapper->addMapping(ui->cameraWidthSpinBox, View::kWviewFieldNumber);
+  viewMapper->addMapping(ui->cameraHeightSpinBox, View::kHviewFieldNumber);
+
+  viewMapper->addMapping(ui->viewportXSpinBox, View::kXportFieldNumber);
+  viewMapper->addMapping(ui->viewportYSpinBox, View::kYportFieldNumber);
+  viewMapper->addMapping(ui->viewportWidthSpinBox, View::kWportFieldNumber);
+  viewMapper->addMapping(ui->viewportHeightSpinBox, View::kHportFieldNumber);
+
+  viewMapper->addMapping(ui->followingHBorderSpinBox, View::kHborderFieldNumber);
+  viewMapper->addMapping(ui->followingVBorderSpinBox, View::kVborderFieldNumber);
+  viewMapper->addMapping(ui->followingHSpeedSpinBox, View::kHspeedFieldNumber);
+  viewMapper->addMapping(ui->followingVSpeedSpinBox, View::kVspeedFieldNumber);
+  viewMapper->toFirst();
+
+  QMenuView* objMenu = new QMenuView(this);
+  TreeSortFilterProxyModel* treeProxy = new TreeSortFilterProxyModel(this);
+  treeProxy->SetFilterType(TreeNode::TypeCase::kObject);
+  treeProxy->setSourceModel(MainWindow::treeModel);
+  objMenu->setModel(treeProxy);
+  ui->objectSelectButton->setMenu(objMenu);
+
+  connect(objMenu, &QMenu::triggered, this, &RoomEditor::SelectedObjectChanged);
+
+  connect(ui->currentViewComboBox, static_cast<void (QComboBox::*)(int)>(&QComboBox::currentIndexChanged),
+          [=](int index) { viewMapper->setCurrentIndex(index); });
+
+  RepeatedProtoModel* m = roomModel->GetRepeatedSubModel(Room::kInstancesFieldNumber);
+  ui->layersAssetsView->setModel(m);
+  connect(ui->layersAssetsView->selectionModel(), &QItemSelectionModel::selectionChanged,

```

Comment:
This code just maps a combobox to a proto enum, right? This seems sufficiently general to be moved elsewhere.

Reply:
Sort of, it maps your selection in the instance list to which instance is displayed in the asset properties view. You know, you select an instance and it shows its properties. Later, I actually want this selection model shared with the room renderer too, like LGM. This setup is a bit like master detail... or just like Qt Designer for instances instead of widgets. It could be moved into a helper yes, but it's not quite needed anywhere else just yet.

That said, it's interesting that you raised this question. I just realized I would like to deal with multiple selections differently. You see, in Qt Creator, if you select multiple widgets then change a property in its property inspector thingy, then it changes it for all selected widgets that have that common property. I think our instances property inspector should work the same way. That makes it easier to set the property for a bunch of instances at once. If you want to set a single instances property, then just select only that instance.